File: BackEnd/Java/springshop_maven/src/main/java/com/pbshop/springshop/analytics/AnalyticsController.java
package com.pbshop.springshop.analytics;

import java.util.Map;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import com.pbshop.springshop.common.api.ApiResponse;

@RestController
@RequestMapping("${pbshop.api.base-path:/api/v1}/analytics")
public class AnalyticsController {

    private final AnalyticsService analyticsService;

    public AnalyticsController(AnalyticsService analyticsService) {
        this.analyticsService = analyticsService;
    }

    @GetMapping("/products/{id}/lowest-ever")
    public ApiResponse<Map<String, Object>> lowestEver(@PathVariable Long id) {
        return ApiResponse.success(analyticsService.lowestEver(id));
    }

    @GetMapping("/products/{id}/unit-price")
    public ApiResponse<Map<String, Object>> unitPrice(@PathVariable Long id) {
        return ApiResponse.success(analyticsService.unitPrice(id));
    }
}

File: BackEnd/Kotlin/kotlin-ktor-gradle-exposeddao-postgresql/src/main/kotlin/com/pbshop/kotlin/ktor/gradle/exposeddao/postgresql/analytics/AnalyticsController.kt
package com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.analytics

import com.pbshop.kotlin.ktor.gradle.exposeddao.postgresql.common.respondStub
import io.ktor.server.application.call
import io.ktor.server.routing.Route
import io.ktor.server.routing.get

class AnalyticsController(
    private val service: AnalyticsService,
) {
    fun Route.register() {
        get("/analytics/products/{id}/lowest-ever") {
            call.respondStub(service.lowestEver(call.pathInt("id")))
        }
        get("/analytics/products/{id}/unit-price") {
            call.respondStub(service.unitPrice(call.pathInt("id")))
        }
    }

    private fun io.ktor.server.application.ApplicationCall.pathInt(name: String): Int = parameters[name]?.toIntOrNull() ?: 0
}
